### PR TITLE
Added RadonEye RD200 Component

### PR DESCRIPTION
**Pull request by jeffeb3**
# What does this implement/fix? 

This implements a new sensor. A bluetooth compatible sensor called the "RadonEye RD200"

This is based off of some community work to get it connected (which has broken due to the changes in BLE): https://community.home-assistant.io/t/radoneye-ble-interface/94962/66

And I used the code in the Air Things Wave Plus pretty heavily. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1857

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
    esp32_ble_tracker:

    ble_client:
      - mac_address: 01:02:03:04:05:06
        id: radon_eye_ble_id

    sensor:
      - platform: radon_eye_rd200
        ble_client_id: radon_eye_ble_id
        update_interval: 5min # default
        radon:
          name: "Radon"
        radon_long_term:
          name: "Radon Long Term"

# and for the tracker:
    logger:
      level: DEBUG # Required for the tracker to show the device

    esp32_ble_tracker:
    radon_eye_ble:


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jeffeb3** [comment]
I had a lot of respect for ESPHome before. It just works so well whenever I try it. 

But using these tools to develop a new component, I have to say I am very very impressed. There is a lot of quality built into these tools. I fought the device a lot today. But I didn't run into any issues where the esphome API was inconstent or surprising. Every error message I got made sense to me (I do have a lot of experience with C++). The build system was always to the point. The tools are quick.

The CI tools I'm getting through now are also great. I use clang-format and clang-tidy in my day job. But the CI setup is very clear and easy to iterate on (the evicence is the rapid iterations I'm doing trying to get the right space in the place).

I understand now why this project is so high quality from community contributions.

Just wanted to say thank you, and Bravo! :100: :tada:.

**Comment by jeffeb3** [comment]
I am assuming you will sqaush and merge this? Or should I squash and rebase/force push?

I'm just thinking it would be nicer without the development churn.

**Comment by patrickstigler** [comment]
Hello @jeffeb3,

since I had the same idea, but wasn't that fast. I tested your module. 

I am running also a new (latest Firmware) RD200. 


Config:

```
esphome:
  name: livingroom-dht
  platform: ESP32
  board: lolin_d32
# Enable logging
logger:
  level: DEBUG # Required for the tracker to show the device

# Enable Home Assistant API
api:
.....
WIFI CONFIG
.....

external_components:
  - source: github://pr#3119
    components: [ radon_eye_ble, radon_eye_rd200 ]   

radon_eye_ble:
esp32_ble_tracker:
ble_client:
  - mac_address: MACADDRESS
    id: radon_eye_ble_id
sensor:
  - platform: radon_eye_rd200
    ble_client_id: radon_eye_ble_id
    update_interval: 5min # default
    radon:
      name: "Radon"
    radon_long_term:
      name: "Radon Long Term"
```

Console output:

```
[23:46:57][I][app:102]: ESPHome version 2022.1.2 compiled on Jan 27 2022, 23:36:09
...
[23:46:57][C][radon_eye_rd200:166]:   Radon 'Radon'
[23:46:57][C][radon_eye_rd200:166]:     Accuracy Decimals: 0
[23:46:57][C][radon_eye_rd200:166]:     Icon: 'mdi:radioactive'
[23:46:57][C][radon_eye_rd200:167]:   Radon Long Term 'Radon Long Term'
[23:46:57][C][radon_eye_rd200:167]:     State Class: 'measurement'
[23:46:57][C][radon_eye_rd200:167]:     Unit of Measurement: 'Bq/m³'
[23:46:57][C][radon_eye_rd200:167]:     Accuracy Decimals: 0
[23:46:57][C][radon_eye_rd200:167]:     Icon: 'mdi:radioactive'
[23:46:57][C][esp32_ble_tracker:706]: BLE Tracker:
[23:46:57][C][esp32_ble_tracker:707]:   Scan Duration: 300 s
[23:46:57][C][esp32_ble_tracker:708]:   Scan Interval: 320.0 ms
[23:46:57][C][esp32_ble_tracker:709]:   Scan Window: 30.0 ms
...
[23:52:12][I][radon_eye_rd200:015]: Connected successfully!
[23:52:13][I][ble_client:161]: Service UUID: 0x1800
[23:52:13][I][ble_client:162]:   start_handle: 0x1  end_handle: 0x7
[23:52:13][I][ble_client:341]:  characteristic 0x2A00, handle 0x3, properties 0xa
[23:52:13][I][ble_client:341]:  characteristic 0x2A01, handle 0x5, properties 0x2
[23:52:13][I][ble_client:341]:  characteristic 0x2A04, handle 0x7, properties 0x2
[23:52:13][I][ble_client:161]: Service UUID: 0x1801
[23:52:13][I][ble_client:162]:   start_handle: 0x8  end_handle: 0x8
[23:52:13][I][ble_client:161]: Service UUID: 00001523-1212-EFDE-1523-785FEABCD123
[23:52:13][I][ble_client:162]:   start_handle: 0x9  end_handle: 0x11
[23:52:13][I][ble_client:161]: Service UUID: 0x180A
[23:52:13][I][ble_client:162]:   start_handle: 0x12  end_handle: 0xffff
[23:52:14][I][ble_client:341]:  characteristic 0x2A29, handle 0x14, properties 0x2
```


Thanks for your work.

As far as I can tell, it's ready to merge.

**Comment by jesserockz** [comment]
@jeffeb3 I have yet to review, but thank you for your kind words.

We do squash and merge so don't worry about your commits.

I will hopefully get to actually reviewing this before the next release, but as you noticed, clang-format and clang-tidy do a lot of the work already =)

**Comment by boc-tothefuture** [comment]
I can also confirm that this worked for me using @patrickstigler instructions.

**Comment by yieldhog** [comment]
Works for me as well, awesome! Despite my RD200 being configured for pci/l, the data is being pulled in through esphome as Bq/m³. Is there an option to pull as pic/l? Thanks!

Update:

Sorry, I see its in your preview documentation now. Great addition!

```
unit_of_measurement: "pCi/L"
      accuracy_decimals: 2
      filters:
        - lambda: return x / 37;
```

**Comment by jeffeb3** [comment]
The docs show an example of converting back to pCi/L. There is already a unit for Bq/m3, so I didn't add one for pCi/L. I am not sure what the difference is. Docs are here:

https://deploy-preview-1857--esphome.netlify.app/components/sensor/radon_eye_ble.html

The magic conversion is here:
```
unit_of_measurement: "pCi/L"
      accuracy_decimals: 2
      filters:
        - lambda: return x / 37;
 ```

Edit: Ha ha.

**Comment by wbste** [comment]
> I can also confirm that this worked for me using @patrickstigler instructions.

Could I get a copy of your YAML?

Having issues with mine (but likely a user-error as I'm new to all this)

**esphome-web-9ed2ec.yaml**

```
esphome:
  name: esphome-web-9ed2ec

esp32:
  board: esp32dev
  framework:
    type: arduino

external_components:
  # use all components from a local folder
  - source:
      type: git
      url: https://github.com/jeffeb3/esphome/
      ref: radoneye
    components: [ radon_eye_ble, radon_eye_rd200, sensor, esp32_ble_tracker, ble_client]

esp32_ble_tracker:

ble_client:
  - mac_address: 44:17:93:61:8D:5A
    id: radon_eye_ble_id

sensor:
  - platform: radon_eye_rd200
    ble_client_id: radon_eye_ble_id
    update_interval: 5min # default
    radon:
      name: "Radon"
      unit_of_measurement: "pCi/L"
      accuracy_decimals: 2
      filters:
        - lambda: return x / 37.0;
    radon_long_term:
      name: "Radon Long Term"
      unit_of_measurement: "pCi/L"
      accuracy_decimals: 2
      filters:
        - lambda: return x / 37.0;

# Enable logging
logger:
  level: DEBUG # Required for the tracker to show the device

# Enable Home Assistant API
api:

ota:


wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
```

**Logging shows the following:**
```

[10:24:38][I][ble_client:085]: Attempting BLE connection to 44:17:93:61:8d:5a
[10:24:38][I][radon_eye_rd200:015]: Connected successfully!
[10:24:41][I][ble_client:161]: Service UUID: 0x1801
[10:24:41][I][ble_client:162]:   start_handle: 0x1  end_handle: 0x5
[10:24:41][I][ble_client:347]:  characteristic 0x2A05, handle 0x3, properties 0x20
[10:24:41][I][ble_client:161]: Service UUID: 0x1800
[10:24:41][I][ble_client:162]:   start_handle: 0x14  end_handle: 0x1c
[10:24:41][I][ble_client:347]:  characteristic 0x2A00, handle 0x16, properties 0x2
[10:24:41][I][ble_client:347]:  characteristic 0x2A01, handle 0x18, properties 0x2
[10:24:41][I][ble_client:347]:  characteristic 0x2AA6, handle 0x1a, properties 0x2
[10:24:41][I][ble_client:161]: Service UUID: 0x1523
[10:24:41][I][ble_client:162]:   start_handle: 0x28  end_handle: 0xffff
[10:24:41][I][ble_client:347]:  characteristic 0x1524, handle 0x2a, properties 0x6
[10:24:41][I][ble_client:347]:  characteristic 0x1525, handle 0x2c, properties 0x12
[10:24:41][I][ble_client:347]:  characteristic 0x1526, handle 0x2f, properties 0x12
[10:24:41][W][radon_eye_rd200:030]: No sensor read characteristic found at service 00001523-1212-EFDE-1523-785FEABCD123 char 00001525-1212-EFDE-1523-785FEABCD123
[10:24:41][D][esp32_ble_tracker:217]: Starting scan...
```

Do I have to change ABCD123 to the actual service found via nRF Connect?

**Comment by jeffeb3** [comment]
It looks like that is not the right device. Are you sure that Mac address is for an rd200?

It is possible there are different versions or something. I only have one to test with.

I didn't make up the ABCD123. That was from the original python code.

**Comment by wbste** [comment]
> It looks like that is not the right device. Are you sure that Mac address is for an rd200?
> 
> It is possible there are different versions or something. I only have one to test with.
> 
> I didn't make up the ABCD123. That was from the original python code.

Yep, double checked the MAC already. I'll try out some other things.

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+void RadonEyeRD200::read_sensors_(uint8_t *value, uint16_t value_len) {
+  if (value_len < 20) {
+    ESP_LOGD(TAG, "Invalid read");
+    return;
+  }
+
+  // Example data
+  // [13:08:47][D][radon_eye_rd200:107]: result bytes: 5010 85EBB940 00000000 00000000 2200 2500 0000
+  ESP_LOGD(TAG, "result bytes: %02X%02X %02X%02X%02X%02X %02X%02X%02X%02X %02X%02X%02X%02X %02X%02X %02X%02X %02X%02X",
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
  ESP_LOGV(TAG, "result bytes: %02X%02X %02X%02X%02X%02X %02X%02X%02X%02X %02X%02X%02X%02X %02X%02X %02X%02X %02X%02X",
```

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  if (is_valid_radon_value_(radon_month)) {
+    ESP_LOGD(TAG, "Radon Long Term based on month");
+    radon_long_term_sensor_->publish_state(radon_month);
+  } else if (is_valid_radon_value_(radon_day)) {
+    ESP_LOGD(TAG, "Radon Long Term based on day");
+    radon_long_term_sensor_->publish_state(radon_day);
+  }
+
+  ESP_LOGD(TAG, "  Measurements (Bq/m³) now: %0.03f, day: %0.03f, month: %0.03f", radon_now, radon_day, radon_month);
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
  ESP_LOGV(TAG, "  Measurements (Bq/m³) now: %0.03f, day: %0.03f, month: %0.03f", radon_now, radon_day, radon_month);
```

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+  ESP_LOGD(TAG, "  Measurements (pCi/L) now: %0.03f, day: %0.03f, month: %0.03f", radon_now / convert_to_bwpm3,
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
  ESP_LOGV(TAG, "  Measurements (pCi/L) now: %0.03f, day: %0.03f, month: %0.03f", radon_now / convert_to_bwpm3,
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  float radon_day = radon_value.number * convert_to_bwpm3;
+
+  radon_value.chars[0] = value[10];
+  radon_value.chars[1] = value[11];
+  radon_value.chars[2] = value[12];
+  radon_value.chars[3] = value[13];
+  float radon_month = radon_value.number * convert_to_bwpm3;
+
+  if (is_valid_radon_value_(radon_month)) {
+    ESP_LOGD(TAG, "Radon Long Term based on month");
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
    ESP_LOGV(TAG, "Radon Long Term based on month");
```

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    radon_long_term_sensor_->publish_state(radon_month);
+  } else if (is_valid_radon_value_(radon_day)) {
+    ESP_LOGD(TAG, "Radon Long Term based on day");
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
    ESP_LOGV(TAG, "Radon Long Term based on day");
```

File: esphome/components/radon_eye_rd200/radon_eye_rd200.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      parent()->set_enabled(true);
+      parent()->connect();
+    } else {
+      ESP_LOGW(TAG, "Connection in progress");
+    }
+  }
+}
+
+void RadonEyeRD200::write_query_message_() {
+  ESP_LOGD(TAG, "writing 0x50 to write service");
```
Comment (jesserockz) [inline esphome/components/radon_eye_rd200/radon_eye_rd200.cpp]:
```suggestion
  ESP_LOGV(TAG, "writing 0x50 to write service");
```

**Review by jesserockz — COMMENTED** [review]
Sorry just noticed a couple more

**Review by jesserockz — APPROVED** [review]

**Comment by patrickstigler** [comment]
> 

@ace7196

Here is my full config - including the other devices on the esp.
Maybe it helps.


```
esphome:
  name: livingroom-dht
  platform: ESP32
  board: lolin_d32

# Enable logging
logger:
  level: DEBUG # Required for the tracker to show the device
  baud_rate: 0

# Enable Home Assistant API
api:

ota:
  password: "OTA PW"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  # Optional manual IP
  manual_ip:
    static_ip: 192.168.192.11
    gateway: 192.168.192.1
    subnet: 255.255.255.0
    
  ap:
    ssid: "livingroom Fallback Hotspot"
    password: "FALLBack PW"

captive_portal:

external_components:
  - source: github://pr#3119
    components: [ radon_eye_ble, radon_eye_rd200 ]
    
radon_eye_ble:
esp32_ble_tracker:
ble_client:
  - mac_address: EA:1B:62:6F:A4:66
    id: radon_eye_ble_id

sensor:
  - platform: dht
    pin: GPIO16
    model: AM2302
    temperature: 
      name: "Living Room Temperature"
      filters:
       - offset: -0.0
    humidity:
      name: "Living Room Humidity"
    update_interval: 60s
    
  - platform: radon_eye_rd200
    ble_client_id: radon_eye_ble_id
    radon:
      name: "Radon"
    radon_long_term:
      name: "Radon Long Term"
    update_interval: 300s
```

**Pull request merged**